client/verta/verta/dataset/_dataset.py

```diff
@@ -93,14 +93,106 @@ def _set_commit_and_blob_path(self, commit, blob_path):
+            elif component_blob.path.startswith(component_path_as_dir):
+                if implicit_download_to_path:
+                    # default to path relative to parent of `component_path`
+                    #     For example:
+                    #     component_blob.path = "coworker/downloads/data/info.csv"
+                    #     component_path      = "coworker/downloads"
+                    #     local_path          =          "downloads/data/info.csv"
+                    local_path = os.path.relpath(
+                        component_blob.path,
+                        pathlib2.Path(component_path).parent,
+                    )
+
+                    # TODO: move following steps outside of loop if less expensive
+                    path_parts = pathlib2.Path(local_path).parts
+
+                    # user probably doesn't want an s3:/ dir
+                    if path_parts[0] == "s3:":
+                        path_parts = path_parts[1:]
+
+                    # avoid collision with existing directory
+                    root_dirname = path_parts[0]
+                    while os.path.exists(root_dirname):
+                        root_dirname = _file_utils.increment_path(root_dirname)
+                    local_path = os.path.join(root_dirname, *path_parts[1:])
+                else:
+                    # rebase from `component_path` onto `download_to_path`
+                    #     For example:
+                    #     component_blob.path = "coworker/downloads/data/info.csv"
+                    #     component_path      = "coworker/downloads"
+                    #     download_to_path    =            "my-data"
+                    #     local_path          =            "my-data/data/info.csv"
+                    local_path = os.path.join(
+                        download_to_path,
+                        os.path.relpath(component_blob.path, component_path),
+                    )
+
+                components_to_download[component_blob.path] = local_path
+
+        if not components_to_download:
+            raise KeyError("no components found for path {}".format(component_path))
+
+        # determine the root path everything is being downloaded to
+        local_paths = map(os.path.relpath, components_to_download.values())  # relative to cwd
+        root_dirnames = set(local_path.split('/')[0] for local_path in local_paths)
+        if len(root_dirnames) == 1:
+            # great, that's our root
+            downloaded_to_path = root_dirnames.pop()
+        else:
+            # everything is relative to cwd, so cwd is our root
+            # NOTE: we shouldn't end up here if `download_to_path` was provided
+            # TODO: handle if a `local_path` was actually at system root or a parent dir
+            downloaded_to_path = os.curdir
+
+        return (components_to_download, downloaded_to_path)
+
     def download(self, component_path, download_to_path=None, chunk_size=32*(10**3)):
```

Comment:
Should we allow a component path to be empty and mean "everything"? I'd say yes.

Reply:
I have VR-4875 for `dataset.download()` to mean "download all"